File: bitcache/graph_memory.py
"""Graph memory: entity-relationship storage + vector retrieval.

Combines vector similarity search with knowledge graph traversal
for multi-hop reasoning. Stores (entity, relation, entity) triples
and links them to vector embeddings.

Based on: "HippoRAG: Neurobiologically Inspired Long-Term Memory
for Large Language Models" (Gutiérrez et al., NeurIPS 2024)

Architecture:
  - Entities stored as nodes with vector embeddings
  - Relations stored as directed edges between entities
  - Search: vector similarity finds entry points, graph traversal
    expands context via relationships
"""

from collections import defaultdict
from typing import Any, Dict, List, Optional, Set, Tuple

import numpy as np

from bitcache.streaming import StreamingIndex


class GraphMemory:
    """Knowledge graph + vector retrieval for multi-hop reasoning.

    Stores entities with embeddings and connects them via typed
    relations. Search combines vector similarity (find relevant
    entities) with graph traversal (expand via relationships).

    Args:
        dim: Vector dimensionality for entity embeddings.
        max_hops: Maximum graph traversal depth during search.
    """

    def __init__(self, dim: int, max_hops: int = 2):
        self.dim = dim
        self.max_hops = max_hops

        self._index = StreamingIndex(dim=dim, rerank_factor=10)

        # Graph structure
        self._edges: Dict[str, List[Tuple[str, str]]] = defaultdict(list)  # entity → [(relation, target)]
        self._reverse_edges: Dict[str, List[Tuple[str, str]]] = defaultdict(list)  # entity → [(relation, source)]
        self._entity_data: Dict[str, Dict[str, Any]] = {}

    def add_entity(
        self,
        entity_id: str,
        vector: np.ndarray,
        name: Optional[str] = None,
        entity_type: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Add an entity with its embedding.

        Args:
            entity_id: Unique identifier for the entity.
            vector: Embedding vector.
            name: Human-readable name.
            entity_type: Type category (person, org, concept, etc.).
            metadata: Additional metadata.

        Returns:
            The entity ID.
        """
        meta = metadata or {}
        meta["name"] = name or entity_id
        meta["entity_type"] = entity_type or "unknown"

        self._index.insert(vector, id=entity_id, metadata=meta)
        self._entity_data[entity_id] = {
            "name": name or entity_id,
            "entity_type": entity_type or "unknown",
            "metadata": metadata or {},
        }
        return entity_id

    def add_relation(self, source: str, relation: str, target: str) -> bool:
        """Add a directed relation between two entities.

        Args:
            source: Source entity ID.
            relation: Relation type (e.g., "works_at", "integrates_with").
            target: Target entity ID.

        Returns:
            True if both entities exist and relation was added.
        """
        if source not in self._entity_data or target not in self._entity_data:
            return False

        # Avoid duplicates
        edge = (relation, target)
        if edge not in self._edges[source]:
            self._edges[source].append(edge)
            self._reverse_edges[target].append((relation, source))
        return True

    def get_relations(self, entity_id: str) -> List[Dict[str, str]]:
        """Get all outgoing relations from an entity.

        Returns:
            List of {relation, target, target_name} dicts.
        """
        results = []
        for relation, target in self._edges.get(entity_id, []):
            target_data = self._entity_data.get(target, {})
            results.append({
                "relation": relation,
                "target": target,
                "target_name": target_data.get("name", target),
            })
        return results

    def get_incoming_relations(self, entity_id: str) -> List[Dict[str, str]]:
        """Get all incoming relations to an entity.

        Returns:
            List of {relation, source, source_name} dicts.
        """
        results = []
        for relation, source in self._reverse_edges.get(entity_id, []):
            source_data = self._entity_data.get(source, {})
            results.append({
                "relation": relation,
                "source": source,
                "source_name": source_data.get("name", source),
            })
        return results

    def search(
        self,
        query: np.ndarray,
        k: int = 5,
        expand: bool = True,
        max_hops: Optional[int] = None,
    ) -> List[Dict[str, Any]]:
        """Search by vector similarity + graph expansion.

        1. Find top entities by vector similarity
        2. Expand each via graph traversal (follow relations)
        3. Return entities with their relationship context

        Args:
            query: Query vector.
            k: Number of seed entities from vector search.
            expand: Whether to expand via graph traversal.
            max_hops: Override default max_hops.

        Returns:
            List of result dicts with entity info and related context.
        """
        scores, ids, metas = self._index.search(query, k=k)

        if not ids:
            return []

        hops = max_hops if max_hops is not None else self.max_hops
        results = []

        for score, entity_id, meta in zip(scores, ids, metas):
            result = {
                "id": entity_id,
                "name": meta.get("name", entity_id),
                "entity_type": meta.get("entity_type", "unknown"),
                "score": float(score),
                "relations": self.get_relations(entity_id),
                "incoming": self.get_incoming_relations(entity_id),
            }

            if expand and hops > 0:
                result["expanded"] = self._expand(entity_id, hops)

            results.append(result)

        return results

    def search_by_relation(
        self,
        source: str,
        relation: Optional[str] = None,
    ) -> List[Dict[str, Any]]:
        """Find entities connected by a specific relation.

        Args:
            source: Source entity ID.
            relation: Filter by relation type. None returns all.

        Returns:
            List of connected entity dicts.
        """
        results = []
        for rel, target in self._edges.get(source, []):
            if relation is not None and rel != relation:
                continue
            target_data = self._entity_data.get(target, {})
            results.append({
                "id": target,
                "name": target_data.get("name", target),
                "entity_type": target_data.get("entity_type", "unknown"),
                "relation": rel,
            })
        return results

    def get_path(self, source: str, target: str, max_depth: int = 3) -> Optional[List[Dict[str, str]]]:
        """Find shortest path between two entities via BFS.

        Args:
            source: Start entity ID.
            target: End entity ID.
            max_depth: Maximum path length.

        Returns:
            List of {entity, relation} steps, or None if no path.
        """
        if source not in self._entity_data or target not in self._entity_data:
            return None

        visited: Set[str] = {source}
        queue: List[Tuple[str, List[Dict[str, str]]]] = [(source, [])]

        while queue:
            current, path = queue.pop(0)

            if len(path) >= max_depth:
                continue

            for relation, neighbor in self._edges.get(current, []):
                if neighbor == target:
                    return path + [{"entity": neighbor, "relation": relation}]

                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append((
                        neighbor,
                        path + [{"entity": current, "relation": relation}],
                    ))

        return None

    def _expand(self, entity_id: str, hops: int) -> List[Dict[str, Any]]:
        """Expand entity context via graph traversal."""
        visited: Set[str] = {entity_id}
        expanded = []
        frontier = [entity_id]

        for hop in range(hops):
            next_frontier = []
            for node in frontier:
                for relation, neighbor in self._edges.get(node, []):
                    if neighbor not in visited:
                        visited.add(neighbor)
                        next_frontier.append(neighbor)
                        neighbor_data = self._entity_data.get(neighbor, {})
                        expanded.append({
                            "id": neighbor,
                            "name": neighbor_data.get("name", neighbor),
                            "relation_from": node,
                            "relation": relation,
                            "hop": hop + 1,
                        })
            frontier = next_frontier

        return expanded

    def remove_entity(self, entity_id: str) -> bool:
        """Remove an entity and all its relations."""
        if entity_id not in self._entity_data:
            return False

        # Remove outgoing edges
        for relation, target in self._edges.get(entity_id, []):
            self._reverse_edges[target] = [
                (r, s) for r, s in self._reverse_edges[target] if s != entity_id
            ]
        del self._edges[entity_id]

        # Remove incoming edges
        for relation, source in self._reverse_edges.get(entity_id, []):
            self._edges[source] = [
                (r, t) for r, t in self._edges[source] if t != entity_id
            ]
        del self._reverse_edges[entity_id]

        del self._entity_data[entity_id]
        self._index.delete(entity_id)
        return True

    @property
    def num_entities(self) -> int:
        return len(self._entity_data)

    @property
    def num_relations(self) -> int:
        return sum(len(edges) for edges in self._edges.values())

File: src/graph_memory.rs
//! Graph memory: entity-relationship storage + vector retrieval.
//!
//! Combines vector similarity search with knowledge graph traversal
//! for multi-hop reasoning. Stores (entity, relation, entity) triples
//! and links them to vector embeddings.

use std::collections::{HashMap, HashSet, VecDeque};

use crate::streaming::StreamingIndex;

/// A relation between two entities.
#[derive(Clone, Debug)]
pub struct Relation {
    pub relation_type: String,
    pub target: String,
}

/// A search result with graph context.
#[derive(Clone, Debug)]
pub struct GraphSearchResult {
    pub id: String,
    pub name: String,
    pub entity_type: String,
    pub score: f32,
    pub relations: Vec<RelationInfo>,
    pub expanded: Vec<ExpandedEntity>,
}

#[derive(Clone, Debug)]
pub struct RelationInfo {
    pub relation: String,
    pub target: String,
    pub target_name: String,
}

#[derive(Clone, Debug)]
pub struct ExpandedEntity {
    pub id: String,
    pub name: String,
    pub relation_from: String,
    pub relation: String,
    pub hop: usize,
}

/// Knowledge graph + vector retrieval for multi-hop reasoning.
pub struct GraphMemory {
    dim: usize,
    max_hops: usize,
    index: StreamingIndex,
    edges: HashMap<String, Vec<(String, String)>>,         // entity → [(relation, target)]
    reverse_edges: HashMap<String, Vec<(String, String)>>, // entity → [(relation, source)]
    entity_data: HashMap<String, EntityData>,
}

#[derive(Clone, Debug)]
struct EntityData {
    name: String,
    entity_type: String,
}

impl GraphMemory {
    pub fn new(dim: usize, max_hops: usize) -> Self {
        Self {
            dim,
            max_hops,
            index: StreamingIndex::new(dim, 10),
            edges: HashMap::new(),
            reverse_edges: HashMap::new(),
            entity_data: HashMap::new(),
        }
    }

    /// Add an entity with its embedding.
    pub fn add_entity(
        &mut self,
        entity_id: &str,
        vector: &[f32],
        name: Option<&str>,
        entity_type: Option<&str>,
    ) -> String {
        let name = name.unwrap_or(entity_id).to_string();
        let etype = entity_type.unwrap_or("unknown").to_string();

        let mut meta = HashMap::new();
        meta.insert("name".to_string(), name.clone());
        meta.insert("entity_type".to_string(), etype.clone());

        self.index.insert(vector, Some(entity_id.to_string()), Some(meta));
        self.entity_data.insert(entity_id.to_string(), EntityData {
            name,
            entity_type: etype,
        });

        entity_id.to_string()
    }

    /// Add a directed relation between two entities.
    pub fn add_relation(&mut self, source: &str, relation: &str, target: &str) -> bool {
        if !self.entity_data.contains_key(source) || !self.entity_data.contains_key(target) {
            return false;
        }

        let edge = (relation.to_string(), target.to_string());
        let edges = self.edges.entry(source.to_string()).or_default();
        if !edges.contains(&edge) {
            edges.push(edge);
            self.reverse_edges
                .entry(target.to_string())
                .or_default()
                .push((relation.to_string(), source.to_string()));
        }
        true
    }

    /// Search by vector similarity + graph expansion.
    pub fn search(
        &self,
        query: &[f32],
        k: usize,
        expand: bool,
        max_hops: Option<usize>,
    ) -> Vec<GraphSearchResult> {
        let (scores, ids, metas) = self.index.search(query, k);

        if ids.is_empty() {
            return vec![];
        }

        let hops = max_hops.unwrap_or(self.max_hops);

        ids.into_iter()
            .zip(scores.into_iter())
            .zip(metas.into_iter())
            .map(|((id, score), meta)| {
                let name = meta.get("name").cloned().unwrap_or_else(|| id.clone());
                let entity_type = meta.get("entity_type").cloned().unwrap_or_else(|| "unknown".to_string());

                let relations = self.get_relations(&id);
                let expanded = if expand && hops > 0 {
                    self.expand(&id, hops)
                } else {
                    vec![]
                };

                GraphSearchResult {
                    id,
                    name,
                    entity_type,
                    score,
                    relations,
                    expanded,
                }
            })
            .collect()
    }

    /// Get all outgoing relations from an entity.
    pub fn get_relations(&self, entity_id: &str) -> Vec<RelationInfo> {
        self.edges.get(entity_id)
            .map(|edges| {
                edges.iter().map(|(rel, target)| {
                    let target_name = self.entity_data.get(target)
                        .map(|d| d.name.clone())
                        .unwrap_or_else(|| target.clone());
                    RelationInfo {
                        relation: rel.clone(),
                        target: target.clone(),
                        target_name,
                    }
                }).collect()
            })
            .unwrap_or_default()
    }

    /// Find shortest path between two entities via BFS.
    pub fn get_path(&self, source: &str, target: &str, max_depth: usize) -> Option<Vec<(String, String)>> {
        if !self.entity_data.contains_key(source) || !self.entity_data.contains_key(target) {
            return None;
        }

        let mut visited: HashSet<String> = HashSet::new();
        visited.insert(source.to_string());

        let mut queue: VecDeque<(String, Vec<(String, String)>)> = VecDeque::new();
        queue.push_back((source.to_string(), vec![]));

        while let Some((current, path)) = queue.pop_front() {
            if path.len() >= max_depth {
                continue;
            }

            if let Some(edges) = self.edges.get(&current) {
                for (relation, neighbor) in edges {
                    if neighbor == target {
                        let mut result = path.clone();
                        result.push((relation.clone(), neighbor.clone()));
                        return Some(result);
                    }

                    if !visited.contains(neighbor) {
                        visited.insert(neighbor.clone());
                        let mut new_path = path.clone();
                        new_path.push((relation.clone(), current.clone()));
                        queue.push_back((neighbor.clone(), new_path));
                    }
                }
            }
        }

        None
    }

    /// Remove an entity and all its relations.
    pub fn remove_entity(&mut self, entity_id: &str) -> bool {
        if !self.entity_data.contains_key(entity_id) {
            return false;
        }

        // Remove outgoing edges
        if let Some(edges) = self.edges.remove(entity_id) {
            for (_, target) in &edges {
                if let Some(rev) = self.reverse_edges.get_mut(target) {
                    rev.retain(|(_, src)| src != entity_id);
                }
            }
        }

        // Remove incoming edges
        if let Some(rev_edges) = self.reverse_edges.remove(entity_id) {
            for (_, source) in &rev_edges {
                if let Some(fwd) = self.edges.get_mut(source) {
                    fwd.retain(|(_, tgt)| tgt != entity_id);
                }
            }
        }

        self.entity_data.remove(entity_id);
        self.index.delete(entity_id);
        true
    }

    fn expand(&self, entity_id: &str, hops: usize) -> Vec<ExpandedEntity> {
        let mut visited: HashSet<String> = HashSet::new();
        visited.insert(entity_id.to_string());
        let mut expanded = Vec::new();
        let mut frontier = vec![entity_id.to_string()];

        for hop in 0..hops {
            let mut next_frontier = Vec::new();
            for node in &frontier {
                if let Some(edges) = self.edges.get(node) {
                    for (relation, neighbor) in edges {
                        if !visited.contains(neighbor) {
                            visited.insert(neighbor.clone());
                            next_frontier.push(neighbor.clone());
                            let name = self.entity_data.get(neighbor)
                                .map(|d| d.name.clone())
                                .unwrap_or_else(|| neighbor.clone());
                            expanded.push(ExpandedEntity {
                                id: neighbor.clone(),
                                name,
                                relation_from: node.clone(),
                                relation: relation.clone(),
                                hop: hop + 1,
                            });
                        }
                    }
                }
            }
            frontier = next_frontier;
        }

        expanded
    }

    pub fn num_entities(&self) -> usize {
        self.entity_data.len()
    }

    pub fn num_relations(&self) -> usize {
        self.edges.values().map(|e| e.len()).sum()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_add_entity_and_relation() {
        let dim = 8;
        let mut gm = GraphMemory::new(dim, 2);

        let v1 = vec![1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0];
        let v2 = vec![0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0];

        gm.add_entity("db", &v1, Some("prod-db-01"), Some("database"));
        gm.add_entity("api", &v2, Some("api-gateway"), Some("service"));

        assert!(gm.add_relation("api", "depends_on", "db"));
        assert_eq!(gm.num_entities(), 2);
        assert_eq!(gm.num_relations(), 1);

        let relations = gm.get_relations("api");
        assert_eq!(relations.len(), 1);
        assert_eq!(relations[0].relation, "depends_on");
        assert_eq!(relations[0].target, "db");
    }

    #[test]
    fn test_search_with_expansion() {
        let dim = 8;
        let mut gm = GraphMemory::new(dim, 2);

        let v1 = vec![1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0];
        let v2 = vec![0.0, 0.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0];
        let v3 = vec![0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0];

        gm.add_entity("a", &v1, Some("Entity A"), Some("type1"));
        gm.add_entity("b", &v2, Some("Entity B"), Some("type2"));
        gm.add_entity("c", &v3, Some("Entity C"), Some("type3"));

        gm.add_relation("a", "links_to", "b");
        gm.add_relation("b", "links_to", "c");

        let results = gm.search(&v1, 1, true, None);
        assert!(!results.is_empty());
        assert_eq!(results[0].id, "a");
    }
}
